### ISSUE TITLE
Completion no longer goes component by component

### ISSUE THREAD

**Issue by riastradh** [ledger/ledger-mode#237]
Suppose I have the following accounts ([example ledger](https://github.com/ledger/ledger-mode/files/4500610/test.txt)):

- Assets:Bank:Deposit:20200419
- Assets:Bank:Balance
- Assets:Reimbursements

It used to be that when I typed `As<TAB>`, it would expand to `Assets:`, and then if I typed `B<TAB>` it would expand to `Assets:Bank:`, and then if I typed `D<TAB>` it would expand to `Assets:Bank:Deposit:`.  This has been trained in my muscle memory for years.

Now when I type `As<TAB>`, nothing happens.  The [documentation](https://github.com/ledger/ledger-mode/blob/4dd9b65184fbdac2499368160e9eee5732788399/doc/ledger-mode.texi#L301-L305) suggests that I do

```
(setq-local tab-always-indent 'complete)
(setq-local completion-cycle-threshold t)
(setq-local ledger-complete-in-steps t)
```

But if I do that, then when I type `As<TAB>`, it immediately expands to `Assets:Bank:` even though `Assets:Reimbursements` is another possible completion.

I don't want to disable cycling -- I do want it to cycle through each expansion of each component in the account hierarchy, so that, e.g., after `Assets:Bank:` it will cycle between `Assets:Bank:Balance` and `Assets:Bank:Deposit:`.  But it shouldn't skip ahead and ignore possible completions like `Assets:Reimbursements` -- and I would prefer that it restore the old behaviour of never completing more than one component at a time even if unambiguous, so that I can reliably hit `<TAB>` to complete at most one component as I did before.

Versions:
- GNU Emacs 26.3
- ledger-mode 6a54eb2b762dd2a5a6ceff54019eeb430562948a

**Comment by purcell** [comment]
I just tried this out in Emacs 26.3 and it worked as you wanted it to. Perhaps there are other completion settings affecting the behaviour. What completion system are you using?

**Comment by riastradh** [comment]
```
% git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
% cd ..
% mkdir build-master
% cd build-master
% time nice -n +10 env -i PATH=/bin:/usr/bin:$HOME/usr/bin cmake ../master
[...]
% time nice -n +10 env -i PATH=/bin:/usr/bin:$HOME/usr/bin make -j12
[...]
% emacs -nw -q

; *scratch* buffer
(add-to-list 'load-path (expand-file-name "."))
(load "ledger-mode")
(defun ledger-mode-hook-setup-completion ()
  (setq-local tab-always-indent 'complete)
  (setq-local completion-cycle-threshold t)
  (setq-local ledger-complete-in-steps t))
(add-hook 'ledger-mode-hook #'ledger-mode-hook-setup-completion)
```

Then I opened test.txt, typed `M-x ledger-mode RET`, navigated to the last line, and hit `<TAB>`.  Result: Completed to `Assets:Bank:`.

**Comment by jabranham** [comment]
> I would prefer that it restore the old behaviour of never completing more than one component at a time even if unambiguous, so that I can reliably hit <TAB> to complete at most one component as I did before.

I think in that case you want ledger-complete-in-steps to be nil.

It will still complete more than one component if it's the only match
(i.e if you only have equity:foo and no other equity:bar accounts) but
it should be close enough.

**Comment by jabranham** [comment]
Closing since haven't heard back in ~2 weeks

**Comment by riastradh** [comment]
You haven't heard back in ~2 weeks because nothing changed in ~2 weeks.  The bug is still there: it still doesn't go component by component like it used to, even if I set ledger-complete-in-steps to nil.

**Comment by simonmichael** [comment]
I have the same problem as @riastradh. I just tried this advice:

> I think in that case you want ledger-complete-in-steps to be nil.

For me at least, this seems to have helped: TAB completes one account name component, TAB again adds a second component, TAB again cycles the second component. I think it's resolved for me, except the naming seems very unintuitive. I would never have guess that ledger-complete-in-steps has to be nil so that TAB... completes in steps. :)

**Comment by simonmichael** [comment]
Well, here's one other behaviour, I don't know if it's new or not: the first time you press TAB, after some non-TAB input, it aligns the amount instead of completing. On next TAB it starts completing.

**Comment by riastradh** [comment]
OK, I guess my report wasn't as clear as I thought, so let me try again.

In the [example ledger](https://github.com/ledger/ledger-mode/files/4500610/test.txt) I gave as a reproducer, it _used_ to be that if you hit `<TAB>` on the last line, ledger-mode would _first_ complete to `Assets:` -- that is, it would go component by component.  The _next_ `<TAB>` would cycle through:

 - `Assets:Bank:`
 - `Assets:Reimbursements

There are two issues:

1. The documentation advises setting ledger-complete-in-steps (among other things) to t in order to restore the old behaviour, but it doesn't.

   Instead, the first `<TAB>` immediately expands to `Assets:Bank:`, and cycles between that and `Assets:Reimbursements`, rather than expanding just to `Assets:`.

   In other words, it jumped ahead a component faster than I expected from the way my fingers have been trained by years of data entry in the old behaviour.

2. Setting ledger-complete-in-steps to nil helps, in that the first `<TAB>` expands to `Assets:` like before, but the _next_ `<TAB>` now expands all the way to `Assets:Bank:Balance` rather than cycling through `Assets:Bank:` and `Assets:Reimbursements`.

   In other words, it _also_ jumped ahead a component faster than I expected -- just at a different position.

What I would like to see is the old behaviour: expand a single component at a time, unconditionally; `<TAB>` never completes more than one component at once.

**Comment by simonmichael** [comment]
Agreed, I can reproduce this now. Here's that same example ledger with variables embedded for easier testing:
```
; -*- tab-always-indent:complete; completion-cycle-threshold:t; ledger-complete-in-steps:t; -*-

2020-01-01 Opening Balances
    Assets:Bank:Balance                       100.00 EUR
    Equity:Opening Balances

2020-02-01 Acme Widgetry GmbH
    Assets:Bank:Deposit:20200201              200.00 EUR
    Income:Salary

2020-03-01 Gear
    Assets:Reimbursements                      50.00 EUR
    Assets:Bank:Balance

2020-04-01 Fnord
    As
```

**Comment by jabranham** [comment]
Thanks. Can you try applying this diff and see if that fixes the issue? It seems like completion can get called again very quickly and so it completes `As` then `Ass` then `Asse` and so on. When it gets to `Assets:` it then goes on to complete the next "step" even though it shouldn't. Caching seems to help.

```diff
diff --git a/ledger-complete.el b/ledger-complete.el
index 204d66a4..b86508ef 100644
--- a/ledger-complete.el
+++ b/ledger-complete.el
@@ -315,7 +315,7 @@ Looks in `ledger-accounts-file' if set, otherwise the current buffer."
       (let ((prefix (buffer-substring-no-properties start end)))
         (list start end
               (if (functionp collection)
-                  (completion-table-dynamic
+                  (completion-table-with-cache
                    (lambda (_)
                      (cl-remove-if (apply-partially 'string= prefix) (funcall collection))))
                 collection)
```

**Comment by jabranham** [comment]
I've tested this out locally and it seems to fix the issue so proposed in #242

**Comment by simonmichael** [comment]
@jabranham, your #242 branch is also working perfectly for me*. Thanks!

* with settings:
```
; -*- tab-always-indent:complete; completion-cycle-threshold:t; ledger-complete-in-steps:t; -*-
```

**Comment by simonmichael** [comment]
PS it still does this thing, but maybe that's fine:

> the first time you press TAB, after some non-TAB input, it aligns the amount instead of completing. On next TAB it starts completing.

**Comment by riastradh** [comment]
Yes, this seems to fix the issue reported here -- now the settings recommended by the documentation behave more as I expected from past versions.

I also observe the behaviour that @simonmichael followed up about concerning realigning amounts, which is unexpected -- I will use C-c C-q if I want to realign -- and it no longer puts a space after the account name once it has reached the end of a unique completion, but these can be filed as separate issues.

**Comment by simonmichael** [comment]
I'll upgrade the "align before autocomplete" behaviour to "quite annoying". As you autocomplete each part of the account it will change the position of the amount, so the next TAB often does a realign when you're trying to autocomplete more. Often trips me up.

**Comment by simonmichael** [comment]
PS so my vote would be to simply remove the auto-align behaviour from the TAB key; it's too much for one key. But as @riastradh says, this is perhaps a separate issue.

**Comment by purcell** [comment]
> my vote would be to simply remove the auto-align behaviour

FWIW the `ledger-post-auto-align` setting controls this. Not to say the current behaviour is perfect, but it can at least be influenced.

**Comment by simonmichael** [comment]
Ah, that's the one. Thanks Steve.
